backend/app/adapter/sqldb/userrole_integration_test.go

```diff
@@ -0,0 +1,181 @@
+			expectedRoles: []role.Role{role.Admin, role.Basic, role.ChangeLogViewer},
+			hasErr:        false,
+		},
+		{
+			name: "nonexistent user",
+			user: entity.User{
+				ID: "0000",
+			},
+			toAdd:         []role.Role{},
+			expectedRoles: []role.Role{},
+			hasErr:        false,
+		},
+	}
+
+	for _, testCase := range testCases {
+		t.Run(testCase.name, func(t *testing.T) {
+			dbtest.AccessTestDB(
+				dbConnector,
+				dbMigrationTool,
+				dbMigrationRoot,
+				dbConfig,
+				func(sqlDB *sql.DB) {
+					userRoleRepo := sqldb.NewUserRoleSQL(sqlDB)
+
+					for _, toAdd := range testCase.toAdd {
+						err := userRoleRepo.AddRole(testCase.user, toAdd)
+
+						if testCase.hasErr {
+							assert.NotEqual(t, nil, err)
+							return
+						}
+					}
+
+					roles, err := userRoleRepo.GetRoles(testCase.user)
+
+					if testCase.hasErr {
+						assert.NotEqual(t, nil, err)
+						return
+					}
+
+					assert.Equal(t, nil, err)
+					assert.Equal(t, testCase.expectedRoles, roles)
+
+					_, _ = sqlDB.Exec(fmt.Sprintf("TRUNCATE TABLE %s", table.UserRole.TableName))
+				})
+		})
+	}
+}
+
+func TestUserRoleSql_DeleteRole(t *testing.T) {
+	testCases := []struct {
+		name          string
+		user          entity.User
+		toAdd         []role.Role
+		toDelete      role.Role
+		expectedRoles []role.Role
+		hasErr        bool
+	}{
+		{
+			name: "should delete the record",
+			user: entity.User{
+				ID: "1343",
+			},
+			toAdd:         []role.Role{role.ChangeLogViewer},
+			toDelete:      role.ChangeLogViewer,
+			expectedRoles: []role.Role{},
+			hasErr:        false,
+		},
+		{
+			name: "should delete the higher role",
+			user: entity.User{
+				ID: "1343",
+			},
+			toAdd:         []role.Role{role.Admin, role.ChangeLogViewer, role.Basic},
+			toDelete:      role.Admin,
+			expectedRoles: []role.Role{role.Basic, role.ChangeLogViewer},
+			hasErr:        false,
+		},
+		{
+			name: "should do nothing",
+			user: entity.User{
+				ID: "1343",
+			},
+			toAdd:         []role.Role{role.Admin, role.Basic, role.ChangeLogViewer},
+			toDelete:      role.ChangeLogEditor,
+			expectedRoles: []role.Role{role.Admin, role.Basic, role.ChangeLogViewer},
+			hasErr:        false,
+		},
+		{
+			name: "should do nothing for nonexistent user",
+			user: entity.User{
+				ID: "1343",
+			},
+			toAdd:         []role.Role{},
+			toDelete:      role.ChangeLogEditor,
+			expectedRoles: []role.Role{},
+			hasErr:        false,
+		},
+	}
+
+	for _, testCase := range testCases {
+		t.Run(testCase.name, func(t *testing.T) {
+			dbtest.AccessTestDB(
+				dbConnector,
+				dbMigrationTool,
+				dbMigrationRoot,
+				dbConfig,
+				func(sqlDB *sql.DB) {
+					userRoleRepo := sqldb.NewUserRoleSQL(sqlDB)
+
+					for _, toAdd := range testCase.toAdd {
+						err := userRoleRepo.AddRole(testCase.user, toAdd)
+
+						if testCase.hasErr {
+							assert.NotEqual(t, nil, err)
+							return
+						}
+					}
```

Comment:
Let's have preset data in the DB instead of manually invoking add role to ensure the test has only one reason to fail. Also, not related after the previous suggested change, here we should assume `AddRole` must succeed because it's not the object being tested. So we could use `assert.Equal(t, nil, err)` directly.